hooks/boot/01-kafo-hook-extensions.rb

```diff
@@ -36,6 +36,19 @@ def apply_puppet_code(code)
     Open3.capture3(*Kafo::PuppetCommand.format_command("echo \"#{code}\" | #{bin_path} apply --detailed-exitcodes"))
   end
 
+  def foreman_maintain(command)
```

Comment:
I don't want to rely on foreman_maintain, as I've made very clear. This makes it way too easy to use it. I think at most we should expose a `stop_all_services(exclude = [])` method. IMHO `start_service` should directly invoke `systemctl`. We don't have any more bulk service start code. We can also expose locking explicitly. That does mean a helper is useful so I'm wondering if we can make it private. AFAIK hooks are executed as an instance eval so they have access to private methods as well, which is slightly ugly.

Reply:
I'm sorry but I do not agree that code by obscurity is the way to prevent further use of foreman-maintain. Relying on it sparingly does not change the fact that its being used and needs to be used at the present moment. We have to tackle the root of the issue which is the overall design in my opinion. And I am working on that.

Reply:
I think that exposing the subcommands rather than a generic command is at least a better design since it abstracts away the exact tool and states the intention. Even if we don't hide `foreman_maintain` now, I think that `stop_all_services(exclude = [])` and `start_service` are good methods to implement.

Reply:
So I started writing this and I realized that I am writing more code that is effectively wrapping the subcommand in a "generic" command that looks just like the subcommand. So all the benefits of it looking and feeling like a generic command to me is lost. I can still go this route but it feels like a lot of overkill when I can just call foreman_maintain rather than hide that I cam calling it. Looking ahead we can drop some of this hook code that calls out to maintain and re-think how we handle these kind of "upgrade" tasks.

Reply:
@ekohl thoughts on my latest response?

Reply:
I still think they make sense because they make it easier to get rid of foreman_maintain.